File: app/routes/health.py
"""GET /api/health, live connectivity check for Claude + Ollama + Chroma + Welsh.

the Next.js frontend proxies this endpoint to fill in the CONNECTION card.
"""
from __future__ import annotations

import asyncio
import logging
import subprocess
from functools import lru_cache
from pathlib import Path

import httpx
from fastapi import APIRouter

from app.config import PROJECT_ROOT, get_settings
from app.services import llm, rag, welsh


@lru_cache(maxsize=1)
def _build_sha() -> str:
    # short git sha so the operator can confirm the running backend is
    # on the expected commit. read once at startup, cached forever.
    try:
        out = subprocess.check_output(
            ["git", "rev-parse", "--short", "HEAD"],
            cwd=str(PROJECT_ROOT),
            stderr=subprocess.DEVNULL,
            timeout=2.0,
        )
        return out.decode("ascii").strip() or "unknown"
    except Exception:
        return "unknown"


router = APIRouter()
log = logging.getLogger("u-pal-py.health")

# these headers help when we're tunnelling through ngrok (skips the
# browser-warning interstitial) and give the request a useful user-agent.
# ref: https://ngrok.com/docs/cloud-edge/modules/browser-warning/
CF_HEADERS = {
    "User-Agent":                 "UPal-UWTSD-Chatbot/1.0",
    "ngrok-skip-browser-warning": "true",
}


async def _probe_ollama(base_url: str) -> str:
    # quick ping against /api/tags. if it 200s, Ollama is alive.
    if not base_url:
        return "not_configured"
    try:
        async with httpx.AsyncClient(timeout=6.0) as client:
            r = await client.get(f"{base_url}/api/tags", headers=CF_HEADERS)
            return "connected" if r.is_success else f"http_{r.status_code}"
    except httpx.TimeoutException:
        return "timeout"
    except Exception:
        return "offline"


async def _probe_anthropic(api_key: str) -> str:
    # we hit GET /v1/models rather than doing a generation call, so the
    # health check doesn't burn tokens every minute.
    # ref: https://docs.claude.com/en/api/models-list
    if not api_key:
        return "not_configured"
    url = "https://api.anthropic.com/v1/models"
    headers = {
        "x-api-key":         api_key,
        "anthropic-version": "2023-06-01",
    }
    try:
        async with httpx.AsyncClient(timeout=6.0) as client:
            r = await client.get(url, headers=headers)
            if r.is_success:
                return "connected"
            if r.status_code in (401, 403):
                return "auth_error"
            return f"http_{r.status_code}"
    except httpx.TimeoutException:
        return "timeout"
    except Exception:
        return "offline"


async def _probe_chroma() -> tuple[str, int]:
    # this is a sync call under the hood so we push it to a worker thread
    # with asyncio.to_thread, otherwise we'd block the event loop.
    try:
        count = await asyncio.to_thread(rag.get_collection_count)
        return ("ready" if count > 0 else "empty"), count
    except Exception as e:
        log.warning("Chroma probe failed: %s", e)
        return "offline", 0


@router.get("/health")
async def health() -> dict:
    s = get_settings()

    # asyncio.gather runs all three probes in parallel instead of one after
    # the other, so the health check returns in ~6 s worst case rather than
    # ~18 s if all three timed out.
    # ref: https://docs.python.org/3/library/asyncio-task.html#asyncio.gather
    anthropic_status, ollama_status, (chroma_status, chroma_count) = await asyncio.gather(
        _probe_anthropic(s.anthropic_api_key),
        _probe_ollama(s.ollama_url),
        _probe_chroma(),
    )

    # figure out which provider the llm router will actually use, so the UI
    # can mark the right row as ACTIVE.
    provider_pref = s.llm_provider.lower()
    if provider_pref == "anthropic" and s.anthropic_api_key:
        active_provider = "anthropic"
    else:
        active_provider = "ollama"

    return {
        "status":          "OK",
        "buildSha":        _build_sha(),
        # generation providers
        "provider":        active_provider,
        "anthropic":       anthropic_status,
        "anthropicModel":  s.anthropic_model if s.anthropic_api_key else None,
        "ollama":          ollama_status,
        "ollamaModel":     s.ollama_model or None,
        # RAG vector store
        "chroma":          chroma_status,
        "chromaDocs":      chroma_count,
        # Welsh detector
        "welsh":           "active" if welsh.is_loaded() else "not_loaded",
        "bilingualTerms":  welsh.bilingual_map_size(),
        "welshVocab":      welsh.vocab_size(),
    }

File: frontend/pages/api/health.js
'use strict';
// GET /api/health. live connectivity check used by the CONNECTION card.
//
// primary path: proxy to the Python backend via CHAT_BACKEND_URL. that
// backend knows the real state of Claude, Ollama, Chroma, and the Welsh
// detector. if the tunnel is down we return a degraded response so the
// frontend still renders something.
// ref: https://nextjs.org/docs/pages/api-reference/functions/next-request

const BACKEND_URL = process.env.CHAT_BACKEND_URL;

// ngrok interstitial skip + a friendly user-agent for our own server logs
const CF_HEADERS = {
  'User-Agent':                 'UPal-UWTSD-Chatbot/1.0',
  'ngrok-skip-browser-warning': 'true',
};

async function fetchBackendHealth() {
  if (!BACKEND_URL) return null;
  try {
    // AbortSignal.timeout gives us a built-in timeout without wiring a
    // full AbortController. 8s is enough for the Python backend's three
    // parallel probes but short enough to not hang the Vercel function.
    // ref: https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/timeout_static
    const r = await fetch(`${BACKEND_URL.replace(/\/$/, '')}/api/health`, {
      method:  'GET',
      signal:  AbortSignal.timeout(8000),
      headers: CF_HEADERS,
    });
    if (!r.ok) return null;
    return await r.json();
  } catch {
    return null;
  }
}

export default async function handler(req, res) {
  if (req.method !== 'GET') return res.status(405).json({ error: 'Method not allowed' });

  const backend = await fetchBackendHealth();

  if (backend) {
    // pass the backend's report through so the UI sees the real state.
    // we whitelist the fields instead of spreading the whole object so
    // a future backend version can't leak internal fields accidentally.
    return res.status(200).json({
      status:          'OK',
      provider:        backend.provider        || 'anthropic',
      anthropic:       backend.anthropic       ?? 'not_configured',
      anthropicModel:  backend.anthropicModel  || null,
      ollama:          backend.ollama          ?? 'not_configured',
      ollamaModel:     backend.ollamaModel     || null,
      chroma:          backend.chroma          ?? 'empty',
      chromaDocs:      backend.chromaDocs      ?? 0,
      welsh:           backend.welsh           || 'active',
      bilingualTerms:  backend.bilingualTerms  ?? 0,
      welshVocab:      backend.welshVocab      ?? 0,
    });
  }

  // backend unreachable. we still return 200 so the CONNECTION card can
  // render (just with everything shown as offline) instead of showing
  // a raw fetch error in the UI.
  return res.status(200).json({
    status:          'DEGRADED',
    provider:        null,
    anthropic:       'offline',
    anthropicModel:  null,
    ollama:          'offline',
    ollamaModel:     null,
    chroma:          'offline',
    chromaDocs:      0,
    welsh:           'active',
    bilingualTerms:  0,
    welshVocab:      0,
  });
}
